Add data binding to code editor

pathartl opened the pull request:
Adds standard Blazor data binding to the code editor.

pathartl committed:
Add data binding to code editor

serdarciplak commented:
I see. Thanks!

But the current PR has some issues.
- Following the `GetValue` and `SetValue` methods, this property should be named as `Value`, not `Contents`.
- Besides the `OnDidChangeModelContent` callback, the value may also get updated when the model changes with a `OnDidChangeModel` callback.
- The `GetValue` method call in the callback should be made only when there is an active binding. If not, these unnecessary interop calls on every value change would have important performance impact especially on server apps.
- The property setter should make a `SetValue` call.
- The property getter should throw and exception when there is no active binding for the user to be aware.

ScarletKuro commented on BlazorMonaco/StandaloneCodeEditor.razor.cs:
```diff
@@ -216,7 +216,14 @@ public override async Task EventCallback(string eventName, string eventJson)
+                case "OnDidChangeModelContent":
+                    await OnDidChangeModelContent.InvokeAsync(JsonSerializer.Deserialize<ModelContentChangedEvent>(eventJson, jsonOptions));
+
+                    Contents = await GetValue();
```
Writing directly to components parameter is very bad practice in Blazor.
All you should do is to push `ContentsChanged`, then the parent component that uses `@bind-Contents` will receive the new value, re-render itself, and in turn cause your child component to be re-rendered with the new Contents (see blazor dev comment https://github.com/dotnet/aspnetcore/discussions/49187#discussioncomment-6372174).
If you directly modifying `Contents`, you are messing up with parent-child communication. Sadly MSFT didn't add analyzer yet for this case https://github.com/dotnet/aspnetcore/issues/44690 (see "Warn when a Razor component writes to one of its parameter properties directly.").

Also,
> The property setter should make a SetValue call.

If you are talking about a `[Parameter]` setter/getter, you shouldn't touch it either. They always should be auto properties, otherwise you are violating the [BL007](https://learn.microsoft.com/en-us/aspnet/core/diagnostics/bl0007?view=aspnetcore-9.0). All the validations should happen on Blazor lifecycles, like `OnParameterSet` which is mentioned in the docs.

pathartl committed:
Switch to Value/ValueChanged, bind OnDidChangeModelContent

- Switch to Value/ValueChanged so bind directive becomes @bind-Value
- Invoke ValueChanged on both OnDidChangeModel and OnDidChangeModelContent
- Handle changes with Value parameter

pathartl committed:
Merge branch 'master' into add-data-binding

pathartl commented:
Both sets of feedback should be addressed now.
